### PR TITLE
fix substring function

### LINKED ISSUE #573: SUBSTRING: Unsupported expression: - 3
**Describe the bug**
```
mysql> SELECT SUBSTRING('1234567890' FROM -3 FOR 3);
ERROR 1105 (HY000): Code: 5, displayText = Unsupported expression: - 3, type: UnaryOp { op: Minus, expr: Value(Number("3", false)) }.
```

expect:
```
+--------------------------------+
| SUBSTRING('1234567890', -3, 3) |
+--------------------------------+
| 890                            |
+--------------------------------+
```

### PR DESCRIPTION
**Pull request by b41sh**
## Summary

add `DataArrayArithmetic::data_array_unary_arithmetic_op` to support unary operation.
add some test case from substring

## Changelog

- Bug Fix
- Improvement

## Related Issues

Fixes #573

## Test Plan

Unit Tests
Stateless Tests

### CODE REVIEW

File: common/datavalues/src/data_type.rs [hunk 1]
```diff
@@ -259,6 +259,22 @@ pub fn numerical_arithmetic_coercion(
     }
 }
 
+#[inline]
+pub fn numerical_singed_coercion(val_type: &DataType) -> Result<DataType> {
```
Comment (sundy-li) [inline common/datavalues/src/data_type.rs:263]: single ?

Comment (b41sh) [reply common/datavalues/src/data_type.rs]: fixed, PTAL @sundy-li

Comment (b41sh) [reply common/datavalues/src/data_type.rs]: `numerical_signed_coercion` convert unsigned data type to signed

**Comment by sundy-li** [comment]
Stateless tests are not passed. Please take look at: https://github.com/datafuselabs/datafuse/blob/master/scripts/ci/ci-stateless-tests-standalone.sh

File: common/datavalues/src/data_array_arithmetic.rs [hunk 2]
```diff
@@ -73,4 +73,28 @@ impl DataArrayArithmetic {
...
+                    _ => value.to_array()?
+                };
+
+                let coercion_type =
+                    super::data_type::numerical_signed_coercion(&value_array.data_type())?;
+                let value_array = data_array_cast(&value_array, &coercion_type)?;
+                arrow_primitive_array_negate!(&value_array, &coercion_type)
+            }
+            // @todo support other unary operation
+            _ => Result::Err(ErrorCodes::BadArguments("Unsupported unary operation"))
```
Comment (BohuTANG) [inline common/datavalues/src/data_array_arithmetic.rs:97]: It's better to hint what's the op is in the error:
ErrorCodes::BadArguments(format!( "Unsupported unary operation: {:?} as argument", op))

Comment (b41sh) [reply common/datavalues/src/data_array_arithmetic.rs]: modfied, PTAL @BohuTANG

**Comment by b41sh** [comment]
> Stateless tests are not passed. Please take look at: https://github.com/datafuselabs/datafuse/blob/master/scripts/ci/ci-stateless-tests-standalone.sh

I have fixed the test. PTAL @sundy-li

**Review by sundy-li — APPROVED** [review]
LGTM

**Pull request merged**